src/src/com/microsoft/aad/adal/TokenCacheAccessor.java

```diff
@@ -0,0 +1,317 @@
+// Copyright (c) Microsoft Corporation.
+// All rights reserved.
+//
+// This code is licensed under the MIT License.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+package com.microsoft.aad.adal;
+
+import java.io.UnsupportedEncodingException;
+import java.security.NoSuchAlgorithmException;
+import java.util.ArrayList;
+import java.util.List;
+
+import com.microsoft.aad.adal.AuthenticationResult.AuthenticationStatus;
+
+/**
+ * Internal class handling the interaction with {@link AcquireTokenSilentHandler} and {@link ITokenCacheStore}. 
+ */
+class TokenCacheAccessor {
+    private static final String TAG = TokenCacheAccessor.class.getSimpleName();
+    
+    private final ITokenCacheStore mTokenCacheStore;
+    private final String mAuthority;
+    
+    TokenCacheAccessor(final ITokenCacheStore tokenCacheStore, final String authority) {
+        if (tokenCacheStore == null) {
+            throw new IllegalArgumentException("tokenCacheStore");
+        }
+        
+        if (StringExtensions.IsNullOrBlank(authority)) {
+            throw new IllegalArgumentException("authority");
+        }
+        
+        mTokenCacheStore = tokenCacheStore;
+        mAuthority = authority;
+    }
+    
+    /**
+     * @return Access token from cache. Could be null if AT does not exist or expired. 
+     * This will be a strict match with the user passed in, could be unique userid, 
+     * displayable id, or null user. 
+     * @throws AuthenticationException 
+     */
+    TokenCacheItem getATFromCache(final String resource, final String clientId, final String user) 
+            throws AuthenticationException {
+        final TokenCacheItem accessTokenItem = getRegularRefreshTokenCacheItem(resource, clientId, user);
+        if (accessTokenItem == null) {
```

Comment:
We don't store AT for MRRT and FRT token cache entry. 
For regular RT token cache entry, the original plan was to stop saving RT along with AT if the RT is MRRT, but in case there are user taking dependency on RT and we don't want to break existing behaviors, we didn't do this. If we get MRRT back, we correctly mark the token stored in both regular token cache entry and MRRT token cache entry.